src/ctapipe/io/dl2_tables_preprocessing.py

```diff
@@ -0,0 +1,474 @@
+"""Module containing classes related to event loading and preprocessing"""
+
+from pathlib import Path
+
+import astropy.units as u
+import numpy as np
+from astropy.coordinates import AltAz, SkyCoord
+from astropy.table import Column, QTable, Table, vstack
+
+try:
+    from pyirf.simulations import SimulatedEventsInfo
+    from pyirf.spectral import (
+        DIFFUSE_FLUX_UNIT,
+        POINT_SOURCE_FLUX_UNIT,
+        PowerLaw,
+        calculate_event_weights,
+    )
+    from pyirf.utils import calculate_source_fov_offset, calculate_theta
+except ModuleNotFoundError:
+    SimulatedEventsInfo = None
+    DIFFUSE_FLUX_UNIT = None
+    POINT_SOURCE_FLUX_UNIT = None
+    PowerLaw = None
+    calculate_event_weights = None
+    calculate_source_fov_offset = None
+    calculate_theta = None
+
+
+from tables import NoSuchNodeError
+from traitlets import default
+
+from ..compat import COPY_IF_NEEDED
+from ..containers import CoordinateFrameType
+from ..coordinates import NominalFrame
+from ..core import Component, QualityQuery
+from ..core.traits import Bool, Dict, List, Tuple, Unicode
+from .tableloader import TableLoader
+
+__all__ = ["DL2EventLoader", "DL2EventPreprocessor", "DL2EventQualityQuery"]
+
+
+class DL2EventQualityQuery(QualityQuery):
+    """
+    Event pre-selection quality criteria for IRF computation with different defaults.
+    """
+
+    quality_criteria = List(
+        Tuple(Unicode(), Unicode()),
+        default_value=[
+            (
+                "multiplicity 4",
+                "np.count_nonzero(HillasReconstructor_telescopes,axis=1) >= 4",
+            ),
+            ("valid classifier", "RandomForestClassifier_is_valid"),
+            ("valid geom reco", "HillasReconstructor_is_valid"),
+            ("valid energy reco", "RandomForestRegressor_is_valid"),
+        ],
+        help=QualityQuery.quality_criteria.help,
+    ).tag(config=True)
+
+
+class DL2EventPreprocessor(Component):
+    """Defines pre-selection cuts and the necessary renaming of columns."""
```

Comment:
Not a problem, just a comment, this is not just generalizing the simulation IRF pre-processing and inter-telescope calibration:   this is really a common component to transform DL2.Event data into DL3.Event data! Therefore it should also be used in   #2727.   In principle in fact, what it looks like is that you really want to do cross-calibration on DL3 data, not DL2, so you use this component to produce that in-memory.  

Perhaps that also implies we should name this better?  Is it really pre-processing DL2 events, or post-processing them into DL3? I would say the latter.

Reply:
> In principle in fact, what it looks like is that you really want to do cross-calibration on DL3 data

If I am not mistaken, in DL3 we don't have information per telescope right? So we want to apply the cross calibration in a subset of DL2 data, selected based on gammaness. The point of this MR is to reuse existing ctapipe functionality in order to merge/filter DL2 tables.